pkg/controller/worker/machine_dependencies.go

```diff
@@ -1,25 +1,180 @@
+}
+
+func (w *workerDelegate) reconcileServerGroups(computeClient osclient.Compute, workerStatus *api.WorkerStatus) (serverGroupDependencySet, error) {
+	serverGroupDepSet := newServerGroupDependencySet(workerStatus.ServerGroupDependencies)
+	for _, pool := range w.worker.Spec.Pools {
+		serverGroupDependencyStatus, err := w.reconcilePoolServerGroup(computeClient, pool, serverGroupDepSet)
+		if err != nil {
+			return serverGroupDepSet, fmt.Errorf("reconciling server groups failed for pool %q: %w", pool.Name, err)
+		}
+		serverGroupDepSet.upsert(serverGroupDependencyStatus)
+	}
+	return serverGroupDepSet, nil
+}
+
+func (w *workerDelegate) reconcilePoolServerGroup(computeClient osclient.Compute, pool extensionsv1alpha1.WorkerPool, set serverGroupDependencySet) (*api.ServerGroupDependency, error) {
+	poolProviderConfig, err := helper.WorkerConfigFromRawExtension(pool.ProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if poolProviderConfig == nil || poolProviderConfig.ServerGroupConfig == nil || poolProviderConfig.ServerGroupConfig.Policy == "" {
+		return nil, nil
+	}
+
+	poolDep := set.getByPoolName(pool.Name)
+	if poolDep != nil {
+		serverGroup, err := computeClient.GetServerGroup(poolDep.ID)
+		if err != nil && !osclient.IsNotFoundError(err) {
+			return nil, err
+		} else if err == nil {
+			if serverGroup.Name == poolDep.Name && (len(serverGroup.Policies) > 0 && serverGroup.Policies[0] == poolProviderConfig.ServerGroupConfig.Policy) {
+				// if the current dependency's spec matches the provider resource, do nothing.
+				return nil, nil
+			}
+		}
+	}
+
+	name, err := generateServerGroupName(w.ClusterTechnicalName(), pool.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate server group name for worker pool %q: %w", pool.Name, err)
+	}
+
+	result, err := computeClient.CreateServerGroup(name, poolProviderConfig.ServerGroupConfig.Policy)
+	if err != nil {
+		return nil, err
+	}
+
+	return &api.ServerGroupDependency{
+		PoolName: pool.Name,
+		ID:       result.ID,
+		Name:     result.Name,
+	}, nil
 }
 
 func (w *workerDelegate) CleanupMachineDependencies(ctx context.Context) error {
-	return nil
+	computeClient, err := w.openstackClient.Compute()
+	if err != nil {
+		return err
+	}
+
+	workerStatus, err := w.decodeWorkerProviderStatus()
+	if err != nil {
+		return err
+	}
+
+	serverGroupDepSet := newServerGroupDependencySet(workerStatus.DeepCopy().ServerGroupDependencies)
+	err = w.cleanupServerGroupDependencies(computeClient, serverGroupDepSet)
+
+	return w.updateMachineDependenciesStatus(ctx, workerStatus, serverGroupDepSet.extract(), err)
+}
+
+// cleanupServerGroupDependencies handles deletion of excess server groups managed by the worker.
+// Cases handled:
+// a) worker is terminating and all server groups have to be deleted
+// b) worker pool is deleted
+// c) worker pool's server group configuration (e.g. policy) changed
+// d) worker pool no longer requires use of server groups
```

Comment:
Is case d) actually possible? I though the servergroup + policy are immutable and can't be changed. I mean anyway the handling is the same.

Reply:
Controller fully supports c) and d) cases although these are prohibited in the validation for now. To support these the internal implementation does need to do a rolling node update by generating new MachineClasses. But generally we can disable the immutability checks at any point in the future and the deletion will work.